### ISSUE TITLE
Spec for Input variables default

### ISSUE THREAD

**Issue by jfsiii** [elastic/package-spec#96]
I was working on the TS type & validation for data streams and have a question about valid values for `default` 

https://github.com/elastic/package-spec/blob/22bc835bcf0480c0973b2c6bbc86df1314e9c0ef/versions/1/data_stream/manifest.spec.yml#L60-L69

I'm having a difficult time parsing this schema (as is the JSON schema to TS converter I used). I think it might mean it can be one of these or an array of them, but I'm not entirely sure.

I'd like to give some examples and see if they should be valid. Please add any other examples I might have missed

 - [ ] no `default` key at all (optional)
 - [ ] `default: ''`
 - [ ] `default: 'some string'`
 - [ ] `default: 1234`
 - [ ] `default: true`
 - [ ] `default: []`
 - [ ] `default: ['']`
 - [ ] `default: ['some', 'strings']`
 - [ ] `default: [12, 34]`
 - [ ] `default: [true, false]`
 - [ ] `default: [3, true, 'strings']`

Also, does the value for `multi` affect `default`? e.g. "`multi` must be `true` for `default` to use `array`"?
https://github.com/elastic/package-spec/blob/22bc835bcf0480c0973b2c6bbc86df1314e9c0ef/versions/1/data_stream/manifest.spec.yml#L42-L47

**Comment by ycombinator** [comment]
> I'd like to give some examples and see if they should be valid. Please add any other examples I might have missed

All your examples look right to me. 

The "problem" with this field, `default`, is that it can pretty much contain any value of any type or no value at all (your first example). Its a bit of a challenge to specify this in the spec.

> Also, does the value for `multi` affect `default`? e.g. "`multi` must be `true` for `default` to use `array`"?

TBH, I'm not a 100% sure of this. Intuitively it does feel like when `multi` is `false`, `default` (if set), should be a scalar value and when `multi` is `true`, `default` (if set), should be an array value. Do you (or @nchaulet maybe?) know if there's any such logic in the UI code?

**Comment by jfsiii** [comment]
> All your examples look right to me.

OK, great. Thanks.

> The "problem" with this field, default, is that it can pretty much contain any value of any type or no value at all (your first example). Its a bit of a challenge to specify this in the spec.

This reminds me of the type for JSON values, except there isn't an object/map/dict in this. Is that an accurate summary? i.e the TS for JSON is roughly

```typescript
export type JsonObject = {[Key in string]?: JsonValue};

export interface JsonArray extends Array<JsonValue> {}

export type JsonValue = string | number | boolean | null | JsonObject | JsonArray;
```

I'll take a shot at it then submit an example soon and see what people think.

**Comment by jfsiii** [comment]
Based on what we discussed above, I created an example object with the various permutations for `default`

<details><summary>example object</summary>

yes, each `type` is `text` even if the `default` value isn't. It's mainly a copy-paste issue, but it is also possible/valid according to current spec as far as I can tell

```json
{
  "title": "some title",
  "release": "beta",
  "streams": [
    {
      "title": "",
      "description": "",
      "input": "",
      "vars": [
        {
          "name": "x",
          "type": "text"
        },
        {
          "name": "x",
          "type": "text",
          "default": ""
        },
        {
          "name": "x",
          "type": "text",
          "default": "some string"
        },
        {
          "name": "x",
          "type": "text",
          "default": 1234
        },
        {
          "name": "x",
          "type": "text",
          "default": true
        },
        {
          "name": "x",
          "type": "text",
          "default": []
        },
        {
          "name": "x",
          "type": "text",
          "default": [
            ""
          ]
        },
        {
          "name": "x",
          "type": "text",
          "default": [
            "some",
            "strings"
          ]
        },
        {
          "name": "x",
          "type": "text",
          "default": [
            12,
            34
          ]
        },
        {
          "name": "x",
          "type": "text",
          "default": [
            3,
            "strings",
            false
          ]
        }
      ]
    }
  ]
}
```
</details>

I then [ran it against the current schema](https://www.jsonschemavalidator.net/s/l7uckVvu) and got 4 errors for the array values

<details><summary>screenshots</summary>

<img width="336" alt="Screen Shot 2020-12-15 at 2 08 38 PM" src="https://user-images.githubusercontent.com/57655/102261166-4ae3bc00-3edf-11eb-941c-e9c4acb825ff.png">

<img width="653" alt="Screen Shot 2020-12-15 at 1 58 20 PM" src="https://user-images.githubusercontent.com/57655/102259883-ab71f980-3edd-11eb-8875-176bfd8a859c.png">

</details>

I believe this is because the current code is specifying 'X or an array of strings' instead of 'X or array of X'

[My shallow understanding](https://cswr.github.io/JsonSchema/spec/multiple_types/#how-restrictions-work-under-multiple-types) is that `types: [A, B]` is effectively `anyOf: [ {type: A}, {type: B} }`. 

<details><summary>I tried the more explicit form</summary>

https://www.jsonschemavalidator.net/s/0aPKfypk

```json
"default": {
  "description": "Default value(s) for variable",
  "anyOf": [
    {
      "type": [
        "string",
        "integer",
        "boolean",
        "null"
      ]
    },
    {
      "type": "array",
      "items": {
        "type": [
          "string",
          "integer",
          "boolean",
          "null"
        ]
      }
    }
  ]
}
```
</details>
and it worked
<img width="529" alt="Screen Shot 2020-12-15 at 1 58 40 PM" src="https://user-images.githubusercontent.com/57655/102259884-ac0a9000-3edd-11eb-9628-76bcff9e780f.png">

I pulled the schema for the value out to its own definition 

```json
    "nameTBD": {
      "type": [
        "string",
        "integer",
        "boolean",
        "null"
      ]
    },
```

then used it like so, and [it worked](https://www.jsonschemavalidator.net/s/Dt3wfqLs)

```json
"anyOf": [
  {
    "$ref": "#/definitions/nameTBD"
  },
  {
    "type": "array",
    "items": {
      "$ref": "#/definitions/nameTBD"
    }
  }
]
```


I don't know if arrays of arrays are allowed, e.g

```json
          "default": [
            [ 3, "strings", false ],
            [ 3, "strings", false ],
            [ 3, "strings", false ],
          ]
```

But the previous examples will fail on that. To make that work, we can just add `"array"` to the end of the definition and [it works](https://www.jsonschemavalidator.net/s/Obz1P8c7)

It seems the decisions/questions are
 1. Is there a bug with the spec? It seems arrays can only be empty or contain strings.
 1. Should we change the spec format from the `types` & `items` "shorthand" to a more explicit `anyOf` or other boolean keyword?

I can submit a PR and we can discuss further there, if that's preferable.

**Comment by ycombinator** [comment]
Thanks for testing this out, @jfsiii. I think we should treat this as a bug in the spec. What you did with `anyOf` seems like the right definition: either accept the following scalar types or an array of the same scalar types. Pulling out the list of scalar types into its own common definition would definitely make maintenance easier, should we decide to add or remove types later. Please go ahead and submit a PR with these changes. Thanks again for finding this issue and the fix for it!

**Comment by ycombinator** [comment]
BTW, I'm on PTO starting today through end of next week. So maybe request a review on your PR from @mtojek as well. If he can't ge to it, I'll get to it when I'm back.

**Comment by jfsiii** [comment]
closed in #98
